### PR TITLE
Driver overhaul

### LINKED ISSUE #14: [BUG] Publishes very large values when losing position estimates
### Description of the bug

When we lose any estimate of position, the driver publishes position values several millions in magnitude, possibly due to incorrect handling of error bits. Error handling should be done for reduced performance and invalid data.

### Steps To Reproduce

1. Setup drone and start seapath.
2. Lose position estimate.
3. See driver output.

### Additional Information

The status bits in the KMBinary message should be used

### PR DESCRIPTION
**Pull request by jorgenfj**
- Cleaned up code and made cleaner
- Added service to reset origin
- Added NavSatFix publisher for origin
- Added Static tf pub for world_fox (SEU) used for foxglove visualization
- Abstracted mutex and socket retrieval related logic away from driver
- Socket now handles parsing from binary to KMBinary format
- Added lambda helper functions in KMBinary struct to retrieve status bits used in diagnostic message
- Diagnostic now uses status bits instead of custom defined error thresholds for warnings
- Changed frequency of driver to 100Hz
- Quaternion now correctly uses rad instead of deg
- Modified lat,lon,height to x,y,z conversion to correct model
- Added dynamic transform for vechicle, vechicle-1, vechicle-2 and seapath

NB! need to rename frame names in asv_setup/launch/tf.launch.py to reflect new frame names
world_frame -> world
seapath_frame -> seapath

Also changed structure of tf tree so world is root.

### CODE REVIEW

File: seapath_ros_driver/include/seapath_ros_driver.hpp [hunk 1]
```diff
@@ -79,139 +50,133 @@ namespace seapath
...
 
       /**
-       * @brief Parses the new data received from the socket
+       * @brief Publishes static transform from world NED to world SEU to use for foxglove visualization.
        *
-       * @param data Variable for storing the new data
-       * @return KMBinaryData - the new data. Is only updated with new data IF the socket is connected. That is done by checking the socket.connected boolean.
        */
-      KMBinaryData parse_kmbinary_data(std::vector<uint8_t> data);
+      void publish_static_tf(const rclcpp::Time& time) const;
```
Comment (alekskl01) [inline seapath_ros_driver/include/seapath_ros_driver.hpp]: maybe rename this function, static tf could be lots of transforms

File: seapath_ros_driver/src/seapath_ros_driver.cpp [hunk 2]
```diff
@@ -12,163 +12,178 @@ namespace seapath
...
+        if(!socket_.socket_connected()){
+            response->success = false;
+            response->message = "Cannot reset origin when socket is disconnected";
+            return;
+        }
+
+        if(timer_){
+        timer_.reset();
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(200)); // Wait for timer to finish callback
```
Comment (alekskl01) [inline seapath_ros_driver/src/seapath_ros_driver.cpp]: This is probably fine, but there might be a cleaner/more robust solution with [std::future](https://cplusplus.com/reference/future/future/)

File: seapath_ros_driver/src/seapath_ros_driver.cpp [hunk 3]
```diff
@@ -184,55 +199,67 @@ namespace seapath
...
+        odom_msg.header.stamp = time;
 
-        odom_msg.header.frame_id = "world_frame";
-        odom_msg.child_frame_id = "seapath_frame";
-        float north = data.latitude;
-        float east = data.longitude;
-        float height = data.ellipsoid_height;
+        odom_msg.header.frame_id = "seapath";
+        odom_msg.child_frame_id = "seapath";
+
```
Comment (alekskl01) [inline seapath_ros_driver/src/seapath_ros_driver.cpp]: make sure this is correct

File: seapath_ros_driver/src/seapath_ros_driver.cpp [hunk 4]
```diff
@@ -184,55 +199,67 @@ namespace seapath
...
+            break;
         }
-    
+        
         diagnostic_array.status.push_back(diagnostic_msg);
         return diagnostic_array;
     }
 
-    nav_msgs::msg::Odometry Driver::get_odometry_message(const KMBinaryData &data)
+    nav_msgs::msg::Odometry Driver::get_odometry_message(const KMBinaryData& data, const rclcpp::Time& time) const
```
Comment (alekskl01) [inline seapath_ros_driver/src/seapath_ros_driver.cpp:240]: What is the motivation for not using the timestamp included in KMBinary?

Comment (jorgenfj) [reply seapath_ros_driver/src/seapath_ros_driver.cpp:240]: The time from the seapath is not synced with the pi and/or xavier. If Theo or someone else manages to sync the seapath with the rest of the system we can start using the timestamp from the KMBinary. The reason we can't do it now, although the time difference between subsequent KMBinary messages would be more accurate, tf2 functions will fail when the time difference is too big, as a result of the time not being synced. So this is the easiest method as of now.

File: seapath_ros_driver/src/seapath_ros_driver.cpp (hunk 3 above)
Comment (jorgenfj) [reply seapath_ros_driver/src/seapath_ros_driver.cpp]: Should be fixed now. Good spot :)

File: seapath_ros_driver/include/seapath_ros_driver.hpp (hunk 1 above)
Comment (jorgenfj) [reply seapath_ros_driver/include/seapath_ros_driver.hpp]: renamed to publish_foxglove_vis_frame

File: seapath_ros_driver/src/seapath_ros_driver.cpp (hunk 2 above)
Comment (jorgenfj) [reply seapath_ros_driver/src/seapath_ros_driver.cpp]: std::future didn't really fit this problem. Fixed with mutex and locks to be more robust

File: seapath_ros_driver/src/seapath_ros_driver.cpp (hunk 3 above)
Comment (jorgenfj) [reply seapath_ros_driver/src/seapath_ros_driver.cpp]: Parent is now world frame, and child is seapath. "child_frame_id: This represents the coordinate frame that is moving with the mobile base, typically called the "base frame". This is the frame of reference attached to the robot or vehicle itself". I set it to seapath, could be base_link as well

**Review by alekskl01 — APPROVED** [review]
LGTM

**Pull request merged**